program-cache/program_cache_layer.cpp

```diff
@@ -0,0 +1,892 @@
+    program_cache_dispatch.clGetContextInfo = tdispatch->clGetContextInfo;
+    program_cache_dispatch.clGetDeviceInfo = tdispatch->clGetDeviceInfo;
+    program_cache_dispatch.clGetPlatformIDs = tdispatch->clGetPlatformIDs;
+    program_cache_dispatch.clGetPlatformInfo = tdispatch->clGetPlatformInfo;
+    program_cache_dispatch.clGetProgramBuildInfo = tdispatch->clGetProgramBuildInfo;
+    program_cache_dispatch.clGetProgramInfo = tdispatch->clGetProgramInfo;
+    program_cache_dispatch.clReleaseDevice = tdispatch->clReleaseDevice;
+    program_cache_dispatch.clReleaseProgram = tdispatch->clReleaseProgram;
+    program_cache_dispatch.clReleaseContext = tdispatch->clReleaseContext;
+    return program_cache_dispatch;
+}
+
+} // namespace
+
+pcl::program_cache_layer::program_cache_layer(const _cl_icd_dispatch* tdispatch)
+    : tdispatch_(tdispatch), program_cache_(init_program_cache_dispatch(tdispatch_))
+{
+    assert(tdispatch_ != nullptr);
+}
+
+cl_program pcl::program_cache_layer::clCreateProgramWithSource(cl_context context,
+                                                               cl_uint count,
+                                                               const char** strings,
+                                                               const size_t* lengths,
+                                                               cl_int* errcode_ret) noexcept
+{
+    if (strings == nullptr)
+    {
+        if (errcode_ret) *errcode_ret = CL_INVALID_ARG_VALUE;
+        return nullptr;
+    }
+    std::stringstream sstream;
+    for (size_t i = 0; i < count; ++i)
+    {
+        if (lengths && lengths[i] > 0)
+        {
+            sstream << std::string_view(strings[i], lengths[i]);
+        }
+        else
+        {
+            sstream << std::string_view(strings[i]);
+        }
+    }
+    program_entry entry(context);
+    entry.source_ = sstream.str();
+
+    const std::lock_guard lock(programs_mutex_);
+    const auto index = next_program_idx_++;
+    program_entries_.emplace(index, std::move(entry));
+    if (errcode_ret) *errcode_ret = CL_SUCCESS;
+    return reinterpret_cast<cl_program>(index);
+}
+
+cl_program pcl::program_cache_layer::clCreateProgramWithIL(cl_context context,
+                                                           const void* il,
+                                                           size_t length,
+                                                           cl_int* errcode_ret) noexcept
+{
+    if (il == nullptr)
+    {
+        if (errcode_ret) *errcode_ret = CL_INVALID_ARG_VALUE;
+        return nullptr;
+    }
+    if (!is_il_program_supported(context))
+    {
+        if (errcode_ret) *errcode_ret = CL_INVALID_OPERATION;
+        return nullptr;
+    }
+
+    std::vector<char> source(length);
+    std::copy_n(static_cast<const char*>(il), length, source.begin());
+    program_entry entry(context);
+    entry.source_ = std::move(source);
```

Comment:
Can't we get move via an X-value? `entry.source_ = std::vector<char>(static_cast<const char*>(il), length);` is one line versus 3.